gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/Orchestrator.java

```diff
@@ -359,6 +363,11 @@ public void submitFlowToDagManager(FlowSpec flowSpec, Dag<JobExecutionPlan> jobE
     try {
       // Send the dag to the DagManager
       this.dagManager.get().addDag(jobExecutionPlanDag, true, true);
+
+      // Delete spec from flow catalog for adhoc executions after persisting it in DagManager
+      if (!flowSpec.getConfig().hasPath(ConfigurationKeys.JOB_SCHEDULE_KEY)) {
+        this.flowCatalog.get().remove(flowSpec.getUri(), new Properties(), false);
+      }
```

Comment:
Would this work for launch events of adhoc flows when a leader goes down? (i.e. need to re-compile the flow from the flow catalog in the dag action change monitor).

Reply:
Yes, it will. If we've gotten to the point where `addDag` persists the `dag` to `dagStateStore`, now its responsibility of `dagManager` to execute it. The multi-active scheduler component ensures persisting to `dagActionStore` so the event will not be reprocessed through `dagActionChangeMonitor`.  The `dagActionChangeMonitor` only commits the offset after submitting a launch event to the `Orchestrator` for compiling and completing the `addDag` call. The new leader's `DagManager` will recover it from `DagStateStore` if the old leader never completes it.

Reply:
great news!  let's consider a code comment to preserve this knowledge